### PR TITLE
Builder2

### PR DESCRIPTION
**Pull request by justinboswell**
Depends on https://github.com/awslabs/aws-crt-builder/pull/6

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -137,6 +137,7 @@ def _build_dependency(self, aws_lib):
             '-DBUILD_SHARED_LIBS=OFF',
             '-DCMAKE_BUILD_TYPE={}'.format(build_type),
             '-DBUILD_TESTING=OFF',
+            '-DCMAKE_EXPORT_COMPILE_COMMANDS=ON',
```
Comment (graebm) [inline setup.py]: Does this change actually fix anything? This cmake command is only used to build the `aws-c-*` libraries. The python module is compiled and linked by setup.py magicsauce, not cmake.

Are you trying to solve the issue described in builder.json, where you the chunk about "don't run clang-tidy until we find a way to generate compile_commands.json."?

File: .github/workflows/linux-ci-build.yml [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+    steps:
+    - name: Checkout Source
+      uses: actions/checkout@v1
+    
+    - name: Build and test
+      run: |
+        export CODEBUILD_SRC_DIR=$(pwd)
+        python3 -c "from urllib.request import urlretrieve; urlretrieve('https://raw.githubusercontent.com/awslabs/aws-c-common/builder2/codebuild/builder.py', 'builder.py')"
+        python3 builder.py build linux-clang-9-linux-x64
```
Comment (graebm) [inline .github/workflows/linux-ci-build.yml]: confused about the difference between:
`linux-ci-action.yml` which does `python3 builder.py run aws-crt-python`
and
`linux-ci-build.yml` which does `python3 builder.py build linux-clang-9-linux-x64`

Why bother having `linux-ci-build.yml`?
Isn't the point of the builder2 experiment to stop using builder.json in repos like aws-crt-python where it only gets in the way?

File: .builder/actions/aws-crt-python.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+            ['{python}', '-m', 'pip', 'install', '--upgrade', 'pip', 'setuptools', 'virtualenv'],
+            ['{python}', '-m', 'virtualenv', 'venv'],
+            [venv_python, '-m', 'pip', 'install', '--upgrade', 'autopep8', 'boto3'],
+            [venv_python, '-m', 'pip', 'install', '.', '--install-option=--verbose', '--install-option=build_ext', '--install-option=--include-dirs{openssl_include}', '--install-option=--library-dirs{openssl_lib}'],
+            [venv_python, '-m', 'unittest', 'discover', '--verbose'],
+            [venv_python, 'aws-common-runtime/aws-c-http/integration-testing/http_client_test.py', venv_python, 'elasticurl.py'],
+            [venv_python, '-m', 'autopep8', '--exit-code', '--diff', '--recursive', 'awscrt', 'test', 'setup.py'],
+        ]
+
+        return Builder.Script(actions, name='aws-crt-python')
```
Comment (graebm) [inline .builder/actions/aws-crt-python.py]: I'm interested to see what the manylinux build would look like, where we're running this 5 times, once per variant of python on the machine

File: .github/workflows/linux-ci-build.yml (hunk 2 above)
Comment (justinboswell) [reply .github/workflows/linux-ci-build.yml]: For comparison. Like I've been saying I would do :)

File: .builder/actions/aws-crt-python.py (hunk 3 above)
Comment (justinboswell) [reply .builder/actions/aws-crt-python.py]: Ok I'll do that next.

File: setup.py (hunk 1 above)
Comment (justinboswell) [reply setup.py]: Oh interesting, you're right. I can undo this. I kinda fired that one into the wind while waiting for a build to run.

File: builder.json [hunk 4]
```diff
@@ -25,34 +25,7 @@
             "variables": {
                 "openssl_include": "=/opt/openssl/include",
                 "openssl_lib": "=/opt/openssl/lib"
-            },
-            "post_build_steps": [
-                ["echo", "------ Python 3.6 ------"],
-                ["/opt/python/cp36-cp36m/bin/python", "-m", "pip", "install", "--upgrade", "--trusted-host", "pypi.org", "--trusted-host", "files.pythonhosted.org", "pip", "setuptools", "boto3", "autopep8"],
-                ["/opt/python/cp36-cp36m/bin/python", "setup.py", "--verbose", "build_ext", "--include-dirs{openssl_include}", "--library-dirs{openssl_lib}", "install"],
-                ["/opt/python/cp36-cp36m/bin/python", "-m", "unittest", "discover", "--verbose"],
-                ["/opt/python/cp37-cp37m/bin/python", "aws-common-runtime/aws-c-http/integration-testing/http_client_test.py", "/opt/python/cp36-cp36m/bin/python", "elasticurl.py"],
```
Comment (graebm) [inline builder.json]: woooo so red!

File: setup.py (hunk 1 above)
Comment (graebm) [reply setup.py]: this change still needs to be removed

File: .builder/actions/manylinux1-ci.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+            python = python_path(version)
+            
+            actions = [
+                [python, '-m', 'pip', 'install', '--upgrade',
+                    "--trusted-host", "pypi.org", "--trusted-host", "files.pythonhosted.org",
+                    "pip", "setuptools", "boto3"],
+                [python, '-m', 'pip', 'install', '.',
+                    '--install-option=--verbose', '--install-option=build_ext', '--install-option=--include-dirs{openssl_include}',
+                    '--install-option=--library-dirs{openssl_lib}'],
+                [python3, 'aws-common-runtime/aws-c-http/integration-testing/http_client_test.py', python, 'elasticurl.py'],
```
[the remark this answers is not in the document]
Comment (justinboswell) [reply .builder/actions/manylinux1-ci.py]: that was precisely my thought. That said, I could generalize the installation into a single job and compose them. I'll do that.

File: .builder/actions/aws_crt_python.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+import Builder
+import os
+import sys
+
+class InstallPythonReqs(Builder.Action):
+    def __init__(self, **kwargs):
+        self.trust_hosts = kwargs.get('trust_hosts', False)
+        self.core = ('pip', 'setuptools', 'virtualenv')
+        self.deps = kwargs.get('deps', [])
+        self.python = kwargs.get('python', sys.executable)
```
Comment (graebm) [inline .builder/actions/aws_crt_python.py]: just curious, why not do it like this?
```def init(self, trust_hosts=False, deps=[], python=sys.executable):```

File: .builder/actions/aws_crt_python.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
+
+import Builder
+import os
+import sys
+
+class InstallPythonReqs(Builder.Action):
+    def __init__(self, **kwargs):
+        self.trust_hosts = kwargs.get('trust_hosts', False)
```
Comment (graebm) [inline .builder/actions/aws_crt_python.py]: is there any need to invoke `super()`?

File: .builder/actions/aws_crt_python.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        trusted_hosts = []
+        # These are necessary for older hosts with out of date certs or pip
+        if self.trust_hosts:
+            trusted_hosts = ['--trusted-host', 'pypi.org', '--trusted-host', 'files.pythonhosted.org']
+
+        # setuptools must be installed before packages that need it, or else it won't be in the
+        # package database in time for the subsequent install and pip fails
+        steps = []
+        for deps in (self.core, self.deps):
+            steps.append([self.python, '-m', 'pip', 'install', '--upgrade', *trusted_hosts, *deps])
```
Comment (graebm) [inline .builder/actions/aws_crt_python.py]: ❤ these comments explaining the voodoo

File: .builder/actions/aws_crt_python.py (hunk 6 above)
Comment (justinboswell) [reply .builder/actions/aws_crt_python.py]: Because I'm stupid. Will fix.

File: .builder/actions/aws_crt_python.py (hunk 7 above)
Comment (justinboswell) [reply .builder/actions/aws_crt_python.py]: Nope.

**Pull request merged**